### PR TITLE
Denoising

### PR DESCRIPTION
**Pull request by EuanPyle**
First version of denoising for Tomo_PP to work with my version of CryoCARE (https://github.com/EuanPyle/cryoCARE_mpido)

Requires: input (global) star file with headers: 
_rlnTomoTomogramHalvesForDenoisingEven
_rlnTomoTomogramHalvesForDenoisingOdd

with location of Even and Odd halves of tomogram under the headers.

This PR is not fully operational yet: instead of running the cryoCARE commands, it simply prints them ready for the user to run them which allows easy debugging. Other than that, it's running exactly as it should. Once we're happy with the general way this is set up, I will finalise the module and make the subprocess commands operational.

### CODE REVIEW

**Comment by EuanPyle** [comment]
Sorry for commit madness, but this feature is complete now :) 

just needs the tomogram splitting function to be built into rln and we're good to go!

File: tomography_preprocessing/denoising/care4relion/_utils.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        raise RuntimeError(e)
+    training_tomograms_star = global_star[training_tomograms_idx]
+    return training_tomograms_star
+    
+def find_tomogram_halves(
+        training_tomograms_star: pd.DataFrame,
+) -> Tuple[List, List]:
+    even_tomos = [f'{rows.rlnTomoTomogramHalvesForDenoisingEven}' for idx,rows in training_tomograms_star.iterrows()]
+    odd_tomos = [f'{rows.rlnTomoTomogramHalvesForDenoisingOdd}' for idx,rows in training_tomograms_star.iterrows()]
+    return even_tomos, odd_tomos
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/_utils.py:44]:
```suggestion
return training_tomograms_star['rlnTomoTomogramHalvesForDenoisingEven'], training_tomograms_star['rlnTomoTomogramHalvesForDenoisingOdd'] 
```

not work here?

File: tomography_preprocessing/denoising/care4relion/_utils.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    tomogram_dir.mkdir(parents=True, exist_ok=True)
+    tilt_series_dir = output_directory / 'tilt_series'
+    tilt_series_dir.mkdir(parents=True, exist_ok=True)
+    return training_dir, tomogram_dir, tilt_series_dir
+    
+def parse_training_tomograms(
+        training_tomograms: str
+) -> List:
+    training_tomograms = training_tomograms.strip().split(':')
+    return training_tomograms
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/_utils.py:25]: in general I'm not a huge fan of this pattern for taking multiple CLI args - what's wrong with taking `List[str]` in the CLI?

File: tomography_preprocessing/denoising/care4relion/care4relion.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+from .train_neural_network import train_neural_network
+from ._utils import *
+from .constants import predict_config_prefix, even_suffix 
+from .._cli import cli
+from ...utils.relion import relion_pipeline_job
+
+console = rich.console.Console(record=True)
+
+@cli.command(name='Care4Relion')
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:19]:
```suggestion
@cli.command(name='cryoCARE')
```

general rename suggestion

File: tomography_preprocessing/denoising/care4relion/care4relion.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    """Denoises tomograms using cryoCARE (Thorsten Wagner version) (https://github.com/thorstenwagner/cryoCARE_mpido)
+    
+    Requires that two tomograms have been generated using the same sample. These can be generated via taking odd/even 
+    frames during Motion Correction (optimal) or by taking odd/even tilts during tomogram reconstruction.
+    The location of these tomograms should be specified in the global star file for all tilt series with the headers: 
+    
+     
+    _rlnTomoTomogramHalvesForDenoisingEven
+    
+      _rlnTomoTomogramHalvesForDenoisingOdd
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:38]: really not a fan of these names, how about

`rlnTomoTomogramHalf1` and `rlnTomoTomogramHalf2` to match postprocess?

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 3 above)
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:19]: I feel like this function is 'doing too many things' - why not split it up into 

`relion_tomo_denoise cryoCARE train` and `relion_tomo_denoise cryoCARE denoise`? :)

File: tomography_preprocessing/denoising/care4relion/care4relion.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    if not tilt_series_star_file.exists():
+        e = 'Could not find tilt series star file'
+        console.log(f'ERROR: {e}')
+        raise RuntimeError(e)    
+    
+    if model_name is None and training_tomograms is None:
+        e = 'Both model name and training tomograms are empty. Either provide a trained network through model name or specify \
+            which tomograms to train on (using rlnTomoName of the tilt series, separated by colons) in training tomograms'
+        console.log(f'ERROR: {e}')
+        raise RuntimeError(e)
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:73]: see above comment about 'doing too much'

File: tomography_preprocessing/denoising/care4relion/care4relion.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      
+    global_star = starfile.read(tilt_series_star_file, always_dict=True)['global']
+    
+    if not 'rlnTomoTomogramHalvesForDenoisingEven' in global_star.columns:
+        e = 'Could not find rlnTomoTomogramHalvesForDenoisingEven in tilt series star file.'
+        console.log(f'ERROR: {e}')
+        raise RuntimeError(e)
+    
+    training_dir, tomogram_dir, tilt_series_dir = \
+        create_denoised_tomograms_dir(output_directory)
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:83]: complete names pls

`create_denoising_directory_structure` or something

File: tomography_preprocessing/denoising/care4relion/care4relion.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    training_tomograms_star = generate_training_tomograms_star(
+        global_star=global_star,
+        training_tomograms=training_tomograms,
+    )
+    
+    even_tomos, odd_tomos = find_tomogram_halves(training_tomograms_star)
+    
+    if model_name is None:
+        console.log('User has not provided path to a previously trained neural network, beginning to train now...')
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/care4relion.py:95]: as above

File: tomography_preprocessing/denoising/care4relion/train_neural_network.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+from pathlib import Path
+
+import pandas as pd
+import starfile
+import rich
+import typer
+import subprocess
+from rich.progress import track
+
+from ._utils import *
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/train_neural_network.py:10]: avoid this like the plague! It's impossible to know what came from where and which names might have been clobbered

File: tomography_preprocessing/denoising/care4relion/constants.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+even_suffix = '_even' #The suffix given to even tomograms during split tomogram generation. This will be removed from the names of the output tomograms
+
+predict_config_prefix = 'predict_config' #Name (minus the suffix) of the predict_config.json file
+
+train_data_config_prefix = 'train_data_config' #Name (minus the suffix) of the train_data_config.json file
+
+model_name = 'denoising_model' #Name of the model to be trained
+
+train_config_prefix = 'train_config'
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/constants.py:9]: is split tomogram generation still involved here? 

constants usually have all caps names, not sure if we've been sticking to that

File: tomography_preprocessing/denoising/care4relion/_utils.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+) -> Dict:
+    train_config_json = json.loads(f'{{"train_data": "{training_dir}", "epochs": 100, "steps_per_epoch": 200, "batch_size": 16, "unet_kern_size": 3, \
+    "unet_n_depth": 3, "unet_n_first": 16, "learning_rate": 0.0004, "model_name": "{model_name}", "path": "{training_dir}"}}')
+    return train_config_json
+
+def generate_predict_json(
+        even_tomos: List,
+        odd_tomos: List,
+	training_dir: Path,
+	model_name: str or Path,
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/_utils.py:67]: not a fan of how the type changes what this arg does in the function - is this complexity necessary?

File: tomography_preprocessing/denoising/care4relion/_utils.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+def parse_training_tomograms(
+        training_tomograms: str
+) -> List:
+    training_tomograms = training_tomograms.strip().split(':')
+    return training_tomograms
+
+def generate_training_tomograms_star(
+        global_star: pd.DataFrame,
+        training_tomograms: List,
+) -> pd.DataFrame:
```
Comment (alisterburt) [inline tomography_preprocessing/denoising/care4relion/_utils.py:30]: general lack of docstrings in this module

**Review by alisterburt — CHANGES_REQUESTED** [review]
taken a first pass here! Excited to get this functionality in but would prefer if this were tidied up a little first :) 

Don't hesitate if you have any questions

**Comment by EuanPyle** [comment]
Cheers, will correct tomorrow!

File: tomography_preprocessing/denoising/care4relion/_utils.py (hunk 1 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/_utils.py:44]: Opps, yep! It needs to output as a list so I'm adding .values.list() at the end of each dataframe too

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 3 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:19]: Thought Care4Relion was funny ;) will change tho!

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 4 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:38]: Me neither, in rln I went for rlnTomoReconstructedEvenTomogram which matches the tomogram generation label rlnTomoReconstructedTomogram

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 6 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:83]: doneee

File: tomography_preprocessing/denoising/care4relion/train_neural_network.py (hunk 8 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/train_neural_network.py:10]: was just laziness from me soz! fixed

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 5 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:73]: as above, will separate training and prediction

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 7 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:95]: as above, will separate training and prediction

File: tomography_preprocessing/denoising/care4relion/_utils.py (hunk 11 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/_utils.py:30]: Added!

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 4 above)
Comment (alisterburt) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:38]: Nice! Slight preference for rlnTomoReconstructedTomogramHalf1, I don't think even/odd is as good a name generally - is that okay with you?

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 3 above)
Comment (alisterburt) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:19]: Ooh maybe 
relion_tomo_denoise cryoCARE:train to match the IMOD TSA naming conventions?

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 4 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:38]: I've gone for Odd/Even as that's the MotionCor convention when using splitsum (adds _EVN or _ODD to file names), even/odd may make it easier to trace back to there? Am ok with either approach really

Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:38]: Ohh yes, forgot about the alternating frames

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 3 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py:19]: yep, doing now

File: tomography_preprocessing/denoising/care4relion/_utils.py (hunk 10 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/_utils.py]: fixed

File: tomography_preprocessing/denoising/care4relion/care4relion.py (hunk 3 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/care4relion.py]: doone!

File: tomography_preprocessing/denoising/care4relion/_utils.py (hunk 2 above)
Comment (EuanPyle) [reply tomography_preprocessing/denoising/care4relion/_utils.py]: not sure how we would pass a python list from relion GUI onto cli and into script? str keeps it easy for now

**Comment by EuanPyle** [comment]
All working nicely (haven't managed to check the training output is correct yet as takes 2 hours on our best machine which isn't available atm, but should be fine) is nice and clean!

File: tomography_preprocessing/denoising/care4relion/_utils.py (hunk 2 above)
Comment (alisterburt) [reply tomography_preprocessing/denoising/care4relion/_utils.py]: you could still take a string in RELION, you would just have to change how that string is turned into the CLI arg in `get_commands`/`getComands` where the whole call is generated... let's leave as is for now though

File: README.md [hunk 12]
```diff
@@ -25,6 +25,9 @@ pip install -e .
 | Tilt-series alignment  | `relion_tomo_align_tilt_series IMOD:fiducials`  |
 | Tilt-series alignment  | `relion_tomo_align_tilt_series IMOD:patch-tracking`  |
 | Tilt-series alignment   | `relion_tomo_align_tilt_series AreTomo`  |
+| Denoising (cryoCARE)   | `relion_tomo_denoise`  |
+
+For CryoCARE denoising, please install this version: https://github.com/EuanPyle/cryoCARE_mpido
```
[the remark this answers is not in the document]
Comment (EuanPyle) [reply README.md:30]: I sent a PR to Thorsten Wagner a few days ago but nothing yet. tbf I didn't email I just sent it! 

I fixed a few bugs (e.g. now lets you place output tomos into a dir that already exists and other such small things), and my version allows users to specify a list as input for tomo prediction: previously you had to place all even tomos in one dir and all odd in another, those dirs had to match 100% and have matching names etc. I didn't think this was a great way so now you just enter a list of tomos (same as other steps in the workflow)

[the remark this answers is not in the document]
Comment (alisterburt) [reply README.md:30]: excellent stuff - I'm happy with a fork as long as the idea is that these changes will eventually be upstreamed and we can depend directly on Thorstens package, don't want to maintain another thing :-)

[the remark this answers is not in the document]
Comment (EuanPyle) [reply README.md:30]: me neither! I'm happy to keep an eye on it until they are upsteamed tho.

**Comment by EuanPyle** [comment]
Got everything working on RELION side (will submit a behemoth PR on that end soon, sorry!) with only very minor bugs (e.g. the progress bars from cryoCARE go into run.err rather than run.out etc.).

I've tried installing my cryoCARE (https://github.com/EuanPyle/cryoCARE_mpido) on a few computers now: it's working ok with one exception that I need to pick your brain on! 

To install cryoCARE you need to have cudnn in the LD_LIBRARY_PATH. To do this, upon installation of my version of cryoCARE, I recommend people run:
conda create -n cryocare -c conda-forge -c anaconda python=3 cudnn=8.1.0

There is no pip install for cudnn as it's not a 100% python package, it has to be installed through conda (tho some will have it pre-installed as I think it contains Cuda libraries). Either way, it has to be on LD_LIBRARY_PATH which I do via:
 export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/; 

Which works fine, but having to remind the user to run that command before running rln sounds like a massive faff...Is there a way we can ensure that library is in the LD_LIBRARY_PATH upon install and/or activation of the environment? Or even upon installation of RELION? Also, can we bundle my version of cryoCARE in the setup.cfg of tomo_PP? I can make a pypi repo.

Thanks!

**Comment by alisterburt** [comment]
excellent stuff! 

Which downstream step depends on cudnn being on the LD_LIBRARY_PATH, do you know? 

Overall I would prefer not to duplicate cryoCARE on pypi, is thorstens version already there? We can install your version from source directly, I think that's better :) Installing from source is supported using pip but having sources as a dependency is not as far as I know so I wouldn't add this to the setup yet

making sure LD_LIBRARY_PATH is set correctly is something that should be done as the shell is setup to use a specific program, like setting the PATH, we would set this in some activation script I guess?

**Comment by EuanPyle** [comment]
cudnn being on the LD_[...] allows the GPUs to be found in my experience. If it's not on the path, cryoCARE doesn't see the GPU and automatically skips to CPU processing. GPU(s) are 100% necessary during denoising model training step, I'm probably going to integrate a step in the cryoCARE repo to actually stop training if the GPU can't be found: even on my nice computer's CPU it would have taken a month. It takes a few hrs on GPU. GPU(s) aren't 100% necessary during tomogram generation (prediction), but it still speeds things up ~x5 fold there.

Thorsten's version is on pypi, yes. Source it is then! How do you envisage the user's installing this stuff anyway? Would we guide them on how to set up the correct Python env on the RELION README? i.e. say install relion, then install tomo_pp, then this version of cryoCARE and activate env before RELION startup?

Ok, I'll have a look at some activation scripts and see what I can rustle up! Will get back to you when it's done.

**Comment by EuanPyle** [comment]
Just adding a few minor changes to improve the functionality here, will keep the commits small

**Comment by EuanPyle** [comment]
All done with the minor changes
